File: geocoding_android/android/src/main/java/com/baseflow/geocoding/GeocodingPlugin.java
package com.baseflow.geocoding;

import android.util.Log;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.flutter.embedding.engine.plugins.FlutterPlugin;

/**
 * Plugin implementation that uses the new {@code io.flutter.embedding} package.
 *
 * <p>Instantiate this in an add to app scenario to gracefully handle activity and context changes.
 */
public final class GeocodingPlugin implements FlutterPlugin {
  private static final String TAG = "GeocodingPlugin";
  @Nullable private MethodCallHandlerImpl methodCallHandler;
  @Nullable private Geocoding geocoding;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
    geocoding = new Geocoding(binding.getApplicationContext());
    methodCallHandler = new MethodCallHandlerImpl(geocoding);
    methodCallHandler.startListening(binding.getBinaryMessenger());
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    if (methodCallHandler == null) {
      Log.wtf(TAG, "Already detached from the engine.");
      return;
    }

    methodCallHandler.stopListening();
    methodCallHandler = null;
    geocoding = null;
  }
}

File: geocoding_android/android/src/main/kotlin/com/baseflow/geocoding/GeocodingPlugin.kt
package com.baseflow.geocoding

import androidx.annotation.VisibleForTesting
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding

/**
 * Platform implementation of the geocoding plugin implemented with the Geocoder SDK.
 */
class GeocodingPlugin : FlutterPlugin {
    @VisibleForTesting
    var proxyApiRegistrar: ProxyApiRegistrar? = null

    override fun onAttachedToEngine(binding: FlutterPluginBinding) {
        proxyApiRegistrar =
            ProxyApiRegistrar(
                binding.binaryMessenger,
                binding.applicationContext
            )
        proxyApiRegistrar!!.setUp()
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        if (proxyApiRegistrar != null) {
            proxyApiRegistrar!!.ignoreCallsToDart = true
            proxyApiRegistrar!!.tearDown()
            proxyApiRegistrar!!.instanceManager.stopFinalizationListener()
            proxyApiRegistrar = null
        }
    }
}
